src/mds/SessionMap.h

```diff
@@ -255,27 +256,69 @@ class Session : public RefCountedObject {
 
 class MDS;
 
-class SessionMap {
+/**
+ * Encapsulate the serialized state associated with SessionMap.  Allows
+ * encode/decode outside of live MDS instance.
+ */
+class SessionMapStore {
+protected:
+    mds_rank_t rank;
 public:
-  MDS *mds;
-private:
   ceph::unordered_map<entity_name_t, Session*> session_map;
-public:
   map<int,xlist<Session*>* > by_state;
+  version_t version;
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::iterator& blp);
+  void dump(Formatter *f) const;
+
+  Session* get_or_add_session(const entity_inst_t& i) {
+    Session *s;
+    if (session_map.count(i.name)) {
+      s = session_map[i.name];
+    } else {
+      s = session_map[i.name] = new Session;
+      s->info.inst = i;
+      s->last_cap_renew = ceph_clock_now(g_ceph_context);
+    }
+
+    return s;
+  }
+
+  uint64_t set_state(Session *session, int s) {
+    if (session->state != s) {
+      session->state = s;
+      session->state_seq++;
+      if (by_state.count(s) == 0)
+	by_state[s] = new xlist<Session*>;
+      by_state[s]->push_back(&session->item_session_list);
+    }
+    return session->state_seq;
+  }
+
+  static void generate_test_instances(list<SessionMapStore*>& ls);
+
+  void reset_state()
+  {
+    session_map.clear();
+    by_state.clear();
+  }
+
+  SessionMapStore() : rank(MDS_RANK_NONE), version(0) {}
```

Comment:
Because SessionMap is constructed in MDS before it attains a rank, it goes through this initial period of being rankless.  That used to be implicit (sessionmap->mds->whoami was negative), and is just made explicit here.